### PR TITLE
feat(type): Add support for interval compound literal

### PR DESCRIPTION
**Pull request by anshuldata**
* Introduced APIs to create Literal of type IntervalCompound and IntervalYearToMonth

### CODE REVIEW

**Review by zeroshade — CHANGES_REQUESTED** [review]

File: expr/interval_year_to_month.go
Comment (anshuldata) [reply expr/interval_year_to_month.go:20]: this structure represents value. proto carries type related information too (e.g. [Precision](https://github.com/substrait-io/substrait/blob/28025cbaa8dc3c65b736d8a68fa7070c465fb494/proto/substrait/algebra.proto#L924)). My intention is to construct value by value related fields.
We have another object to represent type so wanted to keep it separate so that separation of value/type remain clear

File: expr/interval_year_to_month.go
Comment (anshuldata) [reply expr/interval_year_to_month.go]: fixed it, Thanks

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go]: fixed it, Thanks

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go:16]: this structure represents value. proto carries type related information too (e.g. [Precision](https://github.com/substrait-io/substrait/blob/28025cbaa8dc3c65b736d8a68fa7070c465fb494/proto/substrait/algebra.proto#L924)). My intention is to construct value by value related fields.
We have another object to represent type so wanted to keep it separate so that separation of value/type remain clear

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go:48]: My thought was to keep it simple for user (set what is needed) and to not expose internal representation. Following reasons

1. I went to Option Pattern since there are more then two fields (more the fields then chances of mistakes are more) to set here and user will typically will set only some of them (depending on interval). If user needs to set Month they just need to give option "WithIntervalCompoundMonths" without worrying about other fields.

2. If we represent it as publicly exposed fields of "IntervalDatePartsConfig" and directly using it in our object. We would have to live with always taking this as input which would have made it hard to change internal representation in future. For internal representation to change, we would have to either copy these structure internally or made a breaking change.

Hence I felt option pattern is more maintainable and easier to use for user in this case

File: expr/interval_compound_literal.go
Comment (jacques-n) [inline expr/interval_compound_literal.go:44]: not a fan of this. think it should probably be WithMilliseconds, WithMicroseconds, WithNanoseconds.

File: plan/relations.go
Comment (jacques-n) [inline plan/relations.go:588]: it looks like some other changes are in this patch. Maybe it needs to be rebased?

File: expr/interval_compound_literal.go
Comment (jacques-n) [inline expr/interval_compound_literal.go:10]: I think there is too much composition and cuteness going on in this patch.

Let's just write a simple struct with public fields that implements Literal interface. We shouldn't use protoliteral. We shouldn't use private structs inside of something else.

Sometimes the simplest thing is the best. Let's keep it simple.

File: expr/interval_compound_literal.go
Comment (jacques-n) [inline expr/interval_compound_literal.go:20]: Let's just hang these methods off of the literal object. So if I wanted, I could do:

new IntervalCompoundLiteral{}.WithMillis(1).WithSeconds(2)...etc.

File: expr/interval_year_to_month.go
Comment (jacques-n) [inline expr/interval_year_to_month.go:8]: Same feedback as other literal. Let's just create a Literal struct impl that is public and give it utility methods as appropriate.

File: plan/relations.go
Comment (anshuldata) [reply plan/relations.go:588]: This is needed because of proto update to v.54. [This](https://github.com/substrait-io/substrait/commit/bed84ecb6193c22bf2ff83dc3a391ec5a9a3aa68) PR updated these types

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go]: fixed it

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go]: fixed it

File: expr/interval_year_to_month.go
Comment (anshuldata) [reply expr/interval_year_to_month.go]: fixed it

**Comment by anshuldata** [comment]
I updated PR to do away with Option and have separate Literal Impl for IntervalDayToMonth and intervalCompound type. Kindly check @jacques-n @zeroshade

File: expr/interval_compound_literal.go
Comment (zeroshade) [inline expr/interval_compound_literal.go:32]: All of these can be made simpler because you're using a value receiver.

```go
func (m IntervalCompoundLiteral) WithYears(years int32) IntervalCompoundLiteral {
   m.years = years
   return m
}
```

The value receiver is going to create a copy when this is called, so you don't need to explicitly make the copy yourself. Same can be done for the rest of the methods

File: expr/interval_compound_literal.go
Comment (zeroshade) [inline expr/interval_compound_literal.go:190]: we should avoid the numerous copies this will make and just construct the compound literal directly here instead. As written, this is going to create at least 4 copies of this struct which seems kinda silly

File: expr/interval_compound_literal.go
Comment (zeroshade) [inline expr/interval_compound_literal.go:212]: shouldn't we return `visit(m)` ?

File: expr/interval_compound_literal.go
Comment (zeroshade) [inline expr/interval_compound_literal.go:20]: since all of the members are private, how does a user actually view or get the value in this literal? As written, it's impossible for any library to do anything with this literal except convert it to and from the protobuf object which makes it sort of useless unless we either make the members public or create accessor methods. 

This is why I was using the protobuf object directly, so I didn't have to duplicate all of the code to create the accessor methods

File: expr/interval_year_to_month.go
Comment (zeroshade) [inline expr/interval_year_to_month.go:23]: same comment as with IntervalCompoundLiteral

File: types/interval_compound_type.go
Comment (zeroshade) [inline types/interval_compound_type.go:31]: same comment as earlier

File: types/interval_compound_type.go
Comment (zeroshade) [inline types/interval_compound_type.go:20]: should the `typeVariationRef` be a parameter? Or do we need a `WithTypeVariationRef` method?

File: types/interval_year_month_type.go
Comment (zeroshade) [inline types/interval_year_month_type.go:32]: why both?

File: types/precison_timestamp_types.go
Comment (zeroshade) [inline types/precison_timestamp_types.go:25]: why this change?

File: plan/relations.go
Comment (jacques-n) [reply plan/relations.go:588]: Best to do that as a separate patch.

File: types/precison_timestamp_types.go
Comment (jacques-n) [reply types/precison_timestamp_types.go:25]: This was a review comment that I failed to submit on the precision patch. It's unrelated to this and should be done separately.

I felt like the implicit iota didn't clarify that the numbers are meaningful. It seemed too easy for a future editor to change things and not realize the importance. Generally, I don't we should iota unless the number is meaningless (just an artifact of go enum pattern) to our code/business logic. I generally follow a similar pattern across languages. For example, in Java, one should store concrete values if they want them and not try to overload ordinal.

File: expr/interval_compound_literal.go
Comment (jacques-n) [reply expr/interval_compound_literal.go:20]: @anshuldata, my request was to make the fields and structure public. No reason to hide.

Comment (jacques-n) [reply expr/interval_compound_literal.go:20]: Also suggest we name the precision field  as subsecond precision or similar. Make it clear that only relates to that field...

File: expr/interval_year_to_month.go
Comment (anshuldata) [reply expr/interval_year_to_month.go]: Resolving this comment, since I assume (based on other comments) that we agree to not use Proto to construct literal and have a separate struct which implements literal is ok

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go]: Resolving this, since I feel (based on later comments) that we agree to not have proto to represent values

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go]: Resolving this. Since it changed now to use builder pattern and public fields

File: plan/relations.go
Comment (anshuldata) [reply plan/relations.go:588]: Raised [this](https://github.com/substrait-io/substrait-go/pull/48) PR for proto update. Once it is merged, I will rebase this PR.

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go]: fixed it, Thanks

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go]: fixed it

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go:158]: I kept is similar to other literal implementations (e.g. [here](https://github.com/substrait-io/substrait-go/blob/cbd28cb19499af1923484ec82540350528249075/expr/literals.go#L501))

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go]: Fixed both
1. Made fields name Public
2. Renamed "precision" to "SubSecondPrecision"

File: types/interval_compound_type.go
Comment (anshuldata) [reply types/interval_compound_type.go:31]: fixed it

File: types/interval_compound_type.go
Comment (anshuldata) [reply types/interval_compound_type.go:20]: added `WithTypeVariationRef`

File: types/precison_timestamp_types.go
Comment (anshuldata) [reply types/precison_timestamp_types.go]: Raised [this](https://github.com/substrait-io/substrait-go/pull/49) PR to fix it. Removed from this PR

File: expr/interval_year_to_month.go
Comment (anshuldata) [reply expr/interval_year_to_month.go]: fixed it

**Comment by anshuldata** [comment]
I have raised separate PR (#48) for proto update. Once it is merged I will rebase this PR

**Comment by anshuldata** [comment]
rebased PR. It is ready for re-review @jacques-n @zeroshade

**Comment by anshuldata** [comment]
rebased it

File: expr/interval_compound_literal.go
Comment (zeroshade) [inline expr/interval_compound_literal.go:69]: since the members are all public now, do we still need these at all?

File: expr/interval_compound_literal_test.go
Comment (zeroshade) [inline expr/interval_compound_literal_test.go:34]: expand this into a multi-line literal to make it easier to read please?

**Review by zeroshade — APPROVED** [review]

File: expr/interval_compound_literal.go
Comment (anshuldata) [reply expr/interval_compound_literal.go]: yes, not useful now. removed them

File: expr/interval_compound_literal_test.go
Comment (anshuldata) [reply expr/interval_compound_literal_test.go]: since remove With* Methods so this test case isn't needed. Removed this test case

**Comment by anshuldata** [comment]
can you merge this @zeroshade, I don't have write access

**Pull request merged**
